runtime/common/src/xcm_sender.rs

```diff
@@ -16,34 +16,71 @@
 
 //! XCM sender for relay chain.
 
+use frame_support::traits::Get;
 use parity_scale_codec::Encode;
-use runtime_parachains::{configuration, dmp};
-use sp_std::marker::PhantomData;
-use xcm::latest::prelude::*;
+use primitives::v2::Id as ParaId;
+use runtime_parachains::{
+	configuration::{self, HostConfiguration},
+	dmp,
+};
+use sp_std::{marker::PhantomData, prelude::*};
+use xcm::prelude::*;
+use SendError::*;
+
+pub trait PriceForParachainDelivery {
+	fn price_for_parachain_delivery(para: ParaId, message: &Xcm<()>) -> MultiAssets;
+}
+impl PriceForParachainDelivery for () {
+	fn price_for_parachain_delivery(_: ParaId, _: &Xcm<()>) -> MultiAssets {
+		MultiAssets::new()
+	}
+}
+
+pub struct ConstantPrice<T>(sp_std::marker::PhantomData<T>);
+impl<T: Get<MultiAssets>> PriceForParachainDelivery for ConstantPrice<T> {
+	fn price_for_parachain_delivery(_: ParaId, _: &Xcm<()>) -> MultiAssets {
+		T::get()
+	}
+}
 
 /// XCM sender for relay chain. It only sends downward message.
-pub struct ChildParachainRouter<T, W>(PhantomData<(T, W)>);
+pub struct ChildParachainRouter<T, W, P>(PhantomData<(T, W, P)>);
 
-impl<T: configuration::Config + dmp::Config, W: xcm::WrapVersion> SendXcm
-	for ChildParachainRouter<T, W>
+impl<T: configuration::Config + dmp::Config, W: xcm::WrapVersion, P: PriceForParachainDelivery>
+	SendXcm for ChildParachainRouter<T, W, P>
 {
-	fn send_xcm(dest: impl Into<MultiLocation>, msg: Xcm<()>) -> SendResult {
-		let dest = dest.into();
-		match dest {
-			MultiLocation { parents: 0, interior: X1(Parachain(id)) } => {
-				// Downward message passing.
-				let versioned_xcm =
-					W::wrap_version(&dest, msg).map_err(|()| SendError::DestinationUnsupported)?;
-				let config = <configuration::Pallet<T>>::config();
-				<dmp::Pallet<T>>::queue_downward_message(
-					&config,
-					id.into(),
-					versioned_xcm.encode(),
-				)
-				.map_err(Into::<SendError>::into)?;
-				Ok(())
-			},
-			dest => Err(SendError::CannotReachDestination(dest, msg)),
-		}
+	type Ticket = (HostConfiguration<T::BlockNumber>, ParaId, Vec<u8>);
+
+	fn validate(
+		dest: &mut Option<MultiLocation>,
+		msg: &mut Option<Xcm<()>>,
+	) -> SendResult<(HostConfiguration<T::BlockNumber>, ParaId, Vec<u8>)> {
+		let d = dest.take().ok_or(MissingArgument)?;
+		let id = if let MultiLocation { parents: 0, interior: X1(Parachain(id)) } = &d {
+			*id
+		} else {
+			*dest = Some(d);
+			return Err(NotApplicable)
+		};
+
+		// Downward message passing.
+		let xcm = msg.take().ok_or(MissingArgument)?;
+		let config = <configuration::Pallet<T>>::config();
```

Comment:
I keep wondering if this is correct, since it's not guaranteed that `deliver` is going to be called in the same block as `validate`, and if there was a configuration change happening in the blocks between calls to `validate` and `deliver`, then `deliver` will be using the old configuration.

Reply:
It is guaranteed. The `Ticket` is not `Encode` so there's no way to save it in one block and bring it out in a second.